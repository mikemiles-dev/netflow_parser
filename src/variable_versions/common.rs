use crate::protocol::ProtocolTypes;

use nom::bytes::complete::take;
use nom::error::{Error as NomError, ErrorKind};
use nom::number::complete::{be_u128, be_u32};
use nom::Err as NomErr;
use nom::IResult;
use nom_derive::*;
use serde::Serialize;

use std::net::{Ipv4Addr, Ipv6Addr};
use std::time::Duration;

/// Holds our datatypes and values post parsing
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize)]
#[serde(untagged)]
pub enum DataNumber {
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    I32(i32),
}

/// Convert into usize, mainly for serialization purposes
impl DataNumber {
    pub fn get_value(self) -> usize {
        let result: usize = self.into();
        result
    }

    /// Parse bytes into DataNumber Type
    pub fn parse(i: &[u8], field_length: u16, signed: bool) -> IResult<&[u8], DataNumber> {
        match field_length {
            1 if !signed => Ok(u8::parse(i)?).map(|(i, j)| (i, Self::U8(j))),
            2 if !signed => Ok(u16::parse(i)?).map(|(i, j)| (i, Self::U16(j))),
            4 if signed => Ok(i32::parse(i)?).map(|(i, j)| (i, Self::I32(j))),
            4 if !signed => Ok(u32::parse(i)?).map(|(i, j)| (i, Self::U32(j))),
            8 if !signed => Ok(u64::parse(i)?).map(|(i, j)| (i, Self::U64(j))),
            16 if !signed => Ok(u128::parse(i)?).map(|(i, j)| (i, Self::U128(j))),
            _ => Err(NomErr::Error(NomError::new(i, ErrorKind::Fail))),
        }
    }

    pub fn from_field_type(
        remaining: &[u8],
        field_type: FieldDataType,
        field_length: u16,
    ) -> IResult<&[u8], FieldValue> {
        let (remaining, field_value) = match field_type {
            FieldDataType::UnsignedDataNumber => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (i, FieldValue::DataNumber(data_number))
            }
            FieldDataType::SignedDataNumber => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, true)?;
                (i, FieldValue::DataNumber(data_number))
            }
            FieldDataType::String => {
                let (i, taken) = take(field_length)(remaining)?;
                (
                    i,
                    FieldValue::String(String::from_utf8_lossy(taken).to_string()),
                )
            }
            FieldDataType::Ip4Addr => {
                let (i, taken) = be_u32(remaining)?;
                let ip_addr = Ipv4Addr::from(taken);
                (i, FieldValue::Ip4Addr(ip_addr))
            }
            FieldDataType::Ip6Addr => {
                let (i, taken) = be_u128(remaining)?;
                let ip_addr = Ipv6Addr::from(taken);
                (i, FieldValue::Ip6Addr(ip_addr))
            }
            FieldDataType::DurationSeconds => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_secs(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationMillis => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_millis(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationMicros => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_micros(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationNanos => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_nanos(data_number.get_value() as u64)),
                )
            }
            FieldDataType::ProtocolType => {
                let (i, protocol) = ProtocolTypes::parse(remaining)?;
                (i, FieldValue::ProtocolType(protocol))
            }
            FieldDataType::Float64 => {
                let (i, f) = f64::parse(remaining)?;
                (i, FieldValue::Float64(f))
            }
            FieldDataType::Vec => {
                let (i, taken) = take(field_length)(remaining)?;
                (i, FieldValue::Vec(taken.to_vec()))
            }
            FieldDataType::Unknown => {
                let (i, taken) = take(field_length)(remaining)?;
                (i, FieldValue::Vec(taken.to_vec()))
            }
        };
        Ok((remaining, field_value))
    }
}
/// Convert into usize, mainly for serialization purposes
impl From<DataNumber> for usize {
    fn from(val: DataNumber) -> Self {
        match val {
            DataNumber::U8(i) => i as usize,
            DataNumber::U32(i) => i as usize,
            DataNumber::I32(i) => i as usize,
            DataNumber::U16(i) => i as usize,
            DataNumber::U64(i) => i as usize,
            DataNumber::U128(i) => i as usize,
        }
    }
}

/// Holds the post parsed field with its relvant datatype
#[derive(Debug, PartialEq, PartialOrd, Clone, Serialize)]
pub enum FieldValue {
    String(String),
    DataNumber(DataNumber),
    Float64(f64),
    Duration(Duration),
    Ip4Addr(Ipv4Addr),
    Ip6Addr(Ipv6Addr),
    Vec(Vec<u8>),
    ProtocolType(ProtocolTypes),
    Unknown,
}

/// Helps the parser indefiy the data type to parse the field as
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub enum FieldDataType {
    String,
    SignedDataNumber,
    UnsignedDataNumber,
    Float64,
    DurationSeconds,
    DurationMillis,
    DurationMicros,
    DurationNanos,
    Ip4Addr,
    Ip6Addr,
    Vec,
    ProtocolType,
    Unknown,
}
