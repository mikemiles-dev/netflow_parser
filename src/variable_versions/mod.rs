pub mod data_number;
pub mod ipfix;
pub mod ipfix_lookup;
pub mod v9;
pub mod v9_lookup;
