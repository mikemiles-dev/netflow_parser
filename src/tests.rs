#[cfg(test)]
mod base_tests {

    use crate::variable_versions::ipfix::{
        Template as IPFixTemplate, TemplateField as IPFixTemplateField,
    };
    use crate::variable_versions::v9::{
        Template as V9Template, TemplateField as V9TemplateField,
    };
    use crate::NetflowParser;

    use hex;
    use insta::assert_yaml_snapshot;

    #[test]
    fn it_parses_unix_timestamp_correctly() {
        use nom::number::complete::{be_u32, be_u64};
        use std::time::Duration;

        let packet = [5, 0, 6, 7, 8, 9, 0, 1];
        let (remain, secs1) =
            be_u32::<&[u8], nom::error::Error<&[u8]>>(packet.as_slice()).unwrap();
        let (remain, nsecs1) = be_u32::<&[u8], nom::error::Error<&[u8]>>(remain).unwrap();
        assert_eq!(remain, []);

        let time1 = Duration::from_nanos(nsecs1 as u64) + Duration::from_secs(secs1 as u64);

        let (remain, secs_nsecs) =
            be_u64::<&[u8], nom::error::Error<&[u8]>>(packet.as_slice()).unwrap();
        assert_eq!(remain, []);
        let secs2 = (secs_nsecs >> 32) as u32 as u64;
        let nsecs2 = secs_nsecs as u32;

        let time2 = Duration::new(secs2, nsecs2);

        assert_eq!(secs1 as u64, secs2);
        assert_eq!(nsecs1, nsecs2);
        assert_eq!(time1, time2);
    }

    #[test]
    #[cfg(not(feature = "unix_timestamp"))]
    fn it_parses_v5() {
        let packet = [
            0, 5, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    #[cfg(feature = "unix_timestamp")]
    fn it_parses_v5_timestamp() {
        let packet = [
            0, 5, 2, 0, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_creates_error() {
        let packet = [12, 13, 14];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    #[cfg(not(feature = "unix_timestamp"))]
    fn it_parses_v7() {
        let packet = [
            0, 7, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    #[cfg(feature = "unix_timestamp")]
    fn it_parses_v7_timestamp() {
        let packet = [
            0, 7, 2, 0, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9() {
        let packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 12, 9, 2, 3, 4, 9, 9, 9, 8,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_no_data() {
        let packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 4,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_template_and_data_packet() {
        let hex = r#"000900090000cc99664728d80000000100000000000000480400001000080004000c000400160004001500040001000400020004000a0004000e0004003d00010088000100070002000b00020004000100060001003c000100050001000000400401000e00080004000c000400160004001500040001000400020004000a0004000e0004003d0001008800010020000200040001003c0001000500010000004808000010001b0010001c001000160004001500040001000400020004000a0004000e0004003d00010088000100070002000b00020004000100060001003c000100050001000000400801000e001b0010001c001000160004001500040001000400020004000a0004000e0004003d000100880001008b000200040001003c0001000500010001001a01000004000c000200040022000400230001005200100100001d000000000000000101616e7900000000000000000000000000080000cd200300d1ef3b2200fe3497fffeb7686e26064700303200000000000068155338000016050000a418000004bf0000000900000000000000000003e2d401bb0619060026064700303200000000000068155338200300d1ef3b2200fe3497fffeb7686e000016050000a418000000b4000000030000000000000000010301bbe2d406040600000000000000000000000000000000010000000000000000000000000000000100004f2c00004f2c0000003c00000001000000000000000000031f91aaea06140600000000040001009df0fb3dc0a801d700006308000076d500000b7200000018000000000000000000031466ac50061a0400c0a801d79df0fb3d00006308000076d500000e370000002100000000000000000103ac501466061e04005dd10e499df0fb3d00006308000076d6000004bd0000000b00000000000000000003ac501466061e04009df0fb3d5dd10e4900006308000076d6000005b90000000c000000000000000001031466ac50061a0400c0a80125c6fcce190000a9180000a9cd0000027b0000000900000000000000000003a34401bb06190400c6fcce19c0a801250000a9180000a9cd00000108000000060000000000000000010301bba34406150400"#;
        let packet = hex::decode(hex).unwrap();
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_multiple_packets() {
        let v9_packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 12, 9, 2, 3, 4, 9, 9, 9, 8,
        ];
        let v7_packet = [
            0, 7, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
        ];
        let v5_packet = [
            0, 5, 0, 1, 3, 0, 4, 0, 5, 0, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3,
            4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1,
            2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        let ipfix_packet = [
            0, 10, 0, 64, 1, 2, 3, 4, 0, 0, 0, 0, 1, 2, 3, 4, 0, 2, 0, 20, 1, 0, 0, 3, 0, 8, 0,
            4, 0, 12, 0, 4, 0, 2, 0, 4, 1, 0, 0, 28, 1, 2, 3, 4, 1, 2, 3, 3, 1, 2, 3, 2, 0, 2,
            0, 2, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        let mut all = vec![];
        all.extend_from_slice(&v9_packet);
        all.extend_from_slice(&v5_packet);
        all.extend_from_slice(&v7_packet);
        all.extend_from_slice(&v9_packet);
        all.extend_from_slice(&ipfix_packet);
        all.extend_from_slice(&v5_packet);
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&all));
    }

    #[test]
    fn it_parses_v9_many_flows() {
        let packet = [
            0, 9, 0, 3, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 16, 1, 2, 0,
            2, 0, 1, 0, 4, 0, 8, 0, 4, 1, 2, 0, 20, 9, 2, 3, 4, 9, 9, 9, 8, 9, 2, 3, 4, 9, 9,
            9, 8,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_options_template() {
        let packet = [
            0, 9, 0, 2, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 22, 1, 19, 0,
            4, 0, 8, 0, 2, 0, 2, 0, 34, 0, 2, 0, 36, 0, 1, 1, 19, 0, 9, 0, 2, 0, 100, 1,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_data_cached_template() {
        let packet = [
            0, 9, 0, 1, 0, 0, 9, 9, 0, 1, 2, 3, 0, 0, 0, 1, 0, 0, 0, 1, 1, 2, 0, 12, 9, 2, 3,
            4, 9, 9, 9, 8,
        ];
        let fields = vec![
            V9TemplateField {
                field_type_number: 1,
                field_type: crate::variable_versions::v9_lookup::V9Field::InBytes,
                field_length: 4,
            },
            V9TemplateField {
                field_type_number: 8,
                field_type: crate::variable_versions::v9_lookup::V9Field::Ipv4SrcAddr,
                field_length: 4,
            },
        ];
        let template = V9Template {
            field_count: 2,
            template_id: 258,
            fields,
        };
        let mut parser = NetflowParser::default();
        parser.v9_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix() {
        let packet = [
            0, 10, 0, 64, 1, 2, 3, 4, 0, 0, 0, 0, 1, 2, 3, 4, 0, 2, 0, 20, 1, 0, 0, 3, 0, 8, 0,
            4, 0, 12, 0, 4, 0, 2, 0, 4, 1, 0, 0, 28, 1, 2, 3, 4, 1, 2, 3, 3, 1, 2, 3, 2, 0, 2,
            0, 2, 0, 1, 2, 3, 4, 5, 6, 7,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_doesnt_panic_with_invalid_options_ipfix_template() {
        let packet = [
            0, 10, 0, 44, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 3, 0, 28, 1, 4, 0, 13, 0, 1,
            128, 123, 0, 4, 0, 0, 0, 2, 0, 41, 0, 2, 0, 42, 0, 2, 0, 0,
        ];
        NetflowParser::default().parse_bytes(&packet);
    }

    #[test]
    fn it_parses_ipfix_data_cached_template() {
        let packet = [
            0, 10, 0, 26, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 2, 0, 10, 0, 8, 0, 0, 1, 1,
        ];
        let fields = vec![
            IPFixTemplateField {
                field_type_number: 2,
                field_type:
                    crate::variable_versions::ipfix_lookup::IPFixField::PacketDeltaCount,
                field_length: 2,
                enterprise_number: None,
            },
            IPFixTemplateField {
                field_type_number: 8,
                field_type:
                    crate::variable_versions::ipfix_lookup::IPFixField::SourceIpv4address,
                field_length: 4,
                enterprise_number: None,
            },
        ];
        let template = IPFixTemplate {
            field_count: 2,
            template_id: 258,
            fields,
        };
        let mut parser = NetflowParser::default();
        parser.ipfix_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix_with_no_template_fields_raises_error() {
        let packet = [
            0, 10, 0, 26, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 2, 0, 10, 0, 8, 0, 0, 1, 1,
        ];
        let template = IPFixTemplate {
            field_count: 2,
            template_id: 258,
            fields: vec![],
        };
        let mut parser = NetflowParser::default();
        parser.ipfix_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_v9_with_no_template_fields_raises_error() {
        let packet = [
            0, 9, 0, 26, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 2, 0, 10, 0, 8, 0, 0, 1, 1,
        ];
        let template = V9Template {
            field_count: 2,
            template_id: 258,
            fields: vec![],
        };
        let mut parser = NetflowParser::default();
        parser.v9_parser.templates.insert(258, template);
        assert_yaml_snapshot!(parser.parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix_options_template() {
        let packet = [
            0, 10, 0, 44, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 3, 0, 28, 1, 4, 0, 3, 0, 1,
            128, 123, 0, 4, 0, 0, 0, 2, 0, 41, 0, 2, 0, 42, 0, 2, 0, 0,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }

    #[test]
    fn it_parses_ipfix_options_template_with_data() {
        let packet = [
            0, 10, 0, 64, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 3, 0, 28, 1, 4, 0, 3, 0, 1,
            128, 123, 0, 4, 0, 0, 0, 2, 0, 41, 0, 2, 0, 42, 0, 2, 0, 0, 1, 4, 0, 20, 0, 0, 0,
            1, 1, 20, 20, 20, 0, 0, 0, 2, 20, 20, 30, 30,
        ];
        assert_yaml_snapshot!(NetflowParser::default().parse_bytes(&packet));
    }
}
