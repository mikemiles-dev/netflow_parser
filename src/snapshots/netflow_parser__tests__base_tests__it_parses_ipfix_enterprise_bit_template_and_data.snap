---
source: src/tests.rs
expression: result
---
- IPFix:
    header:
      version: 10
      length: 244
      export_time: 1530614917
      sequence_number: 34503
      observation_domain_id: 768
    flowsets:
      - header:
          header_id: 2
          length: 228
        body:
          template:
            template_id: 268
            field_count: 34
            fields:
              - field_type_number: 12236
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 12237
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 12240
                field_type: Enterprise
                field_length: 2
                enterprise_number: 9
              - field_type_number: 12241
                field_type: Enterprise
                field_length: 2
                enterprise_number: 9
              - field_type_number: 234
                field_type: IngressVrfid
                field_length: 4
              - field_type_number: 10
                field_type: IngressInterface
                field_length: 4
              - field_type_number: 239
                field_type: BiflowDirection
                field_length: 1
              - field_type_number: 12242
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 95
                field_type: ApplicationId
                field_length: 4
              - field_type_number: 14
                field_type: EgressInterface
                field_length: 4
              - field_type_number: 48
                field_type: SamplerId
                field_length: 1
              - field_type_number: 9252
                field_type: Enterprise
                field_length: 1
                enterprise_number: 9
              - field_type_number: 9253
                field_type: Enterprise
                field_length: 1
                enterprise_number: 9
              - field_type_number: 9357
                field_type: Enterprise
                field_length: 65535
                enterprise_number: 9
              - field_type_number: 12235
                field_type: Enterprise
                field_length: 65535
                enterprise_number: 9
              - field_type_number: 22
                field_type: FlowStartSysUpTime
                field_length: 4
              - field_type_number: 21
                field_type: FlowEndSysUpTime
                field_length: 4
              - field_type_number: 278
                field_type: NewConnectionDeltaCount
                field_length: 4
              - field_type_number: 232
                field_type: ResponderOctets
                field_length: 8
              - field_type_number: 299
                field_type: ResponderPackets
                field_length: 8
              - field_type_number: 231
                field_type: InitiatorOctets
                field_length: 8
              - field_type_number: 298
                field_type: InitiatorPackets
                field_length: 8
              - field_type_number: 9303
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9292
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9300
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9319
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9316
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9268
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9313
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9306
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9307
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9309
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9273
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
              - field_type_number: 9272
                field_type: Enterprise
                field_length: 4
                enterprise_number: 9
- IPFix:
    header:
      version: 10
      length: 188
      export_time: 1530614908
      sequence_number: 34502
      observation_domain_id: 768
    flowsets:
      - header:
          header_id: 268
          length: 172
        body:
          data:
            data_fields:
              - 0:
                  - Enterprise
                  - Vec:
                      - 10
                      - 209
                      - 101
                      - 66
              - 1:
                  - Enterprise
                  - Vec:
                      - 10
                      - 0
                      - 0
                      - 241
              - 2:
                  - Enterprise
                  - Vec:
                      - 192
                      - 224
              - 3:
                  - Enterprise
                  - Vec:
                      - 0
                      - 80
              - 4:
                  - IngressVrfid
                  - DataNumber: 0
              - 5:
                  - IngressInterface
                  - DataNumber: 13
              - 6:
                  - BiflowDirection
                  - DataNumber: 1
              - 7:
                  - Enterprise
                  - Vec:
                      - 141
                      - 195
                      - 63
                      - 224
              - 8:
                  - ApplicationId
                  - String: "\u0003\u0000\u0000P"
              - 9:
                  - EgressInterface
                  - DataNumber: 10
              - 10:
                  - SamplerId
                  - DataNumber: 9
              - 11:
                  - Enterprise
                  - Vec:
                      - 16
              - 12:
                  - Enterprise
                  - Vec:
                      - 0
              - 13:
                  - Enterprise
                  - Vec:
                      - 47
                      - 68
                      - 111
                      - 99
                      - 76
                      - 105
                      - 98
                      - 49
                      - 0
                      - 0
                      - 1
              - 14:
                  - Enterprise
                  - Vec:
                      - 3
                      - 0
                      - 0
                      - 80
                      - 52
                      - 2
                      - 112
                      - 111
                      - 114
                      - 116
                      - 97
                      - 108
                      - 46
                      - 109
                      - 101
                      - 100
                      - 115
                      - 105
                      - 103
                      - 114
                      - 111
                      - 117
                      - 112
                      - 46
                      - 114
                      - 117
              - 15:
                  - FlowStartSysUpTime
                  - DataNumber: 564191867
              - 16:
                  - FlowEndSysUpTime
                  - DataNumber: 564303806
              - 17:
                  - NewConnectionDeltaCount
                  - DataNumber: 1
              - 18:
                  - ResponderOctets
                  - DataNumber: 0
              - 19:
                  - ResponderPackets
                  - DataNumber: 0
              - 20:
                  - InitiatorOctets
                  - DataNumber: 6208
              - 21:
                  - InitiatorPackets
                  - DataNumber: 15
              - 22:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 23:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 24:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 25:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 26:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 27:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 28:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 29:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 30:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 31:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 32:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
              - 33:
                  - Enterprise
                  - Vec:
                      - 0
                      - 0
                      - 0
                      - 0
